iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp

```diff
@@ -168,39 +166,96 @@ struct FlattenBindingSubspan final
+      // fully static.
+      dynamicDim = rewriter.create<ConstantIndexOp>(subspanOp.getLoc(),
+                                                    oldType.getNumElements());
+    } else {
+      ArrayRef<int64_t> oldShape = oldType.getShape();
+      Location loc = subspanOp.getLoc();
+
+      AffineExpr sym0, sym1;
+      bindSymbols(rewriter.getContext(), sym0, sym1);
+      MLIRContext *context = rewriter.getContext();
+      auto mulMap = AffineMap::get(0, 2, {sym0 * sym1}, context);
+
+      Value totalSize = rewriter.create<ConstantIndexOp>(loc, 1);
+      int dynamicDimIndex = 0;
+      for (int i = 0; i < oldType.getRank(); ++i) {
+        Value thisDim;
+        if (ShapedType::isDynamic(oldShape[i])) {
+          thisDim = subspanOp.dynamic_dims()[dynamicDimIndex++];
+        } else {
+          thisDim = rewriter.create<ConstantIndexOp>(loc, oldShape[i]);
+        }
+        totalSize = rewriter.create<AffineApplyOp>(
+            loc, mulMap, ValueRange{totalSize, thisDim});
+      }
+      dynamicDim = totalSize;
+    }
+
     Type newType = getTypeConverter()->convertType(oldType);
 
     rewriter.replaceOpWithNewOp<IREE::HAL::InterfaceBindingSubspanOp>(
         subspanOp, newType, subspanOp.binding(), subspanOp.byte_offset(),
-        subspanOp.byte_length());
+        subspanOp.byte_length(), dynamicDim);
     return success();
   }
 };
 
 /// Generates IR to perform index linearization with the given `indices`
-/// indexing into the given memref `sourceType`.
-static Value linearizeIndices(MemRefType sourceType, ValueRange indices,
+/// indexing into the given memref `sourceValue`.
+static Value linearizeIndices(Value sourceValue, ValueRange indices,
                               Location loc, OpBuilder &builder) {
+  MemRefType sourceType = sourceValue.getType().cast<MemRefType>();
   assert(sourceType.hasRank() && sourceType.getRank() != 0);
-
-  int64_t offset;
-  SmallVector<int64_t, 4> strides;
-  if (failed(getStridesAndOffset(sourceType, strides, offset)) ||
-      llvm::is_contained(strides, MemRefType::getDynamicStrideOrOffset()) ||
-      offset == MemRefType::getDynamicStrideOrOffset()) {
-    return nullptr;
-  }
+  int64_t rank = sourceType.getRank();
 
   AffineExpr sym0, sym1, sym2;
   bindSymbols(builder.getContext(), sym0, sym1, sym2);
   MLIRContext *context = builder.getContext();
   auto mulAddMap = AffineMap::get(0, 3, {sym0 * sym1 + sym2}, context);
 
-  Value linearIndex = builder.create<ConstantIndexOp>(loc, offset);
-  for (auto pair : llvm::zip(indices, strides)) {
-    Value stride = builder.create<ConstantIndexOp>(loc, std::get<1>(pair));
+  int64_t offset;
+  SmallVector<int64_t, 4> strides;
+  strides.reserve(rank);
+
+  // First try to get the strides from the MemRef type itself. This applies to
+  // cases where we have static shapes and only the leading dimension is
+  // dynamic.
+  if (succeeded(getStridesAndOffset(sourceType, strides, offset)) &&
```

Comment:
Could just use `getStridedLayoutMap` (https://github.com/llvm/llvm-project/blob/3419e85b15e35e4b5b4d0cd8153fdbe0958e80c7/mlir/include/mlir/IR/BuiltinTypes.h#L407)

Reply:
Cool! Didn't know about that utility function before.